File: core/java/src/main/java/hara/lang/protocol/IIterator.java
package hara.lang.protocol;

import hara.lang.declaration.HaraMethod;
import hara.lang.declaration.HaraProtocolBinding;
import java.util.Iterator;

/** Stateful traversal cursor. */
@HaraProtocolBinding(
    namespace = "std.protocol.iiterator", name = "IIterator", parents = {"IIter"})
public interface IIterator<E> extends IIter<E>, Iterator<E> {
  @Override
  default Iterator<E> iter() {
    return this;
  }

  @HaraMethod(value = "iter-next?", arity = 1)
  boolean hasNext();

  @HaraMethod(value = "iter-next", arity = 1)
  E next();
}

File: core/rust/src/lang/protocol/iiterator.rs
use super::IIter;
use hara_protocol_macros::hara_protocol;

#[hara_protocol(
    namespace = "std.protocol.iiterator",
    name = "IIterator",
    parents = ["IIter"]
)]
pub trait IIterator: IIter + Iterator<Item = <Self as IIter>::Item> {
    #[hara_method(value = "iter-next", arity = 1)]
    fn iter_next(&mut self) -> Option<<Self as IIter>::Item> {
        self.next()
    }

    #[hara_method(value = "iter-next?", arity = 1)]
    fn iter_next_available(&mut self) -> bool;
}
